src/api.ts

```diff
@@ -0,0 +1,253 @@
+	reject: (error: Error) => void
+
+	processed: boolean
+	processedTime?: number
+	sentResolve: (sent: SendResult) => void
+	// sentReject: (error: Error) => void
+}
+
+export interface Response {
+	reqId?: string
+	command: Commands
+	responseCode: number
+	data: any[]
+
+	type: ResponseTypes
+	message: string
+}
+
+export type ConnectionEvents = {
+	connect: []
+	disconnect: []
+	error: [error: Error]
+}
+
+export class BasicCasparCGAPI extends EventEmitter<ConnectionEvents> {
+	private _connection: Connection
+	private _host: string
+	private _port: number
+
+	private readonly _useSequential: boolean
+	private _requestQueue: Array<InternalRequest> = []
+	private _timeoutTimer: NodeJS.Timer
+	private _timeoutTime: number
+
+	constructor(options?: Options) {
+		super()
+
+		this._host = options?.host || '127.0.0.1'
+		this._port = options?.port || 5250
+		this._useSequential = options?.useSequential || false
+
+		this._connection = new Connection(this._host, this._port)
+
+		this._connection.on('connect', () => {
+			this.emit('connect')
+			this._processQueue().catch((e) => this.emit('error', e))
+		})
+		this._connection.on('disconnect', () => this.emit('disconnect'))
+
+		this._connection.on('data', (response) => {
+			if (this._useSequential) {
+				const request = this._requestQueue.shift()
+
+				if (request) {
+					request.resolve(response)
+				}
+			} else {
+				const request = this._requestQueue.find((req) => req.requestId === response.reqId)
+
+				if (request) {
+					request.resolve(response)
+					this._requestQueue = this._requestQueue.filter((req) => req.requestId !== response.reqId)
+				}
+			}
+
+			this._processQueue().catch(() => null)
+		})
+
+		this._timeoutTime = options?.timeoutTime || 5000
+		this._timeoutTimer = setInterval(() => this._checkTimeouts(), this._timeoutTime)
+	}
+
+	get host(): string {
+		return this._host
+	}
+
+	set host(host: string) {
+		this._host = host
+		this._connection.changeConnection(this._host, this._port)
+	}
+
+	get port(): number {
+		return this._port
+	}
+
+	set port(port: number) {
+		this._port = port
+		this._connection.changeConnection(this._host, this._port)
+	}
+
+	/** Stops internal timers so that the class is ready for garbage disposal */
+	discard(): void {
+		clearInterval(this._timeoutTimer)
+	}
+
+	executeCommand(command: AMCPCommand): Request {
+		const reqId = Math.random().toString(35).slice(2, 7)
+
+		let outerResolve: InternalRequest['sentResolve'] = () => null
+		// outerReject: InternalRequest['sentReject'] = () => null
+		const s = new Promise<SendResult>((resolve) => {
+			outerResolve = resolve
+			// outerReject = reject
+		})
+
+		const internalRequest: InternalRequest = {
+			requestId: reqId,
+			command,
+
+			// stubs to be replaced
+			resolve: () => null,
+			reject: () => null,
+
+			processed: false,
+			sentResolve: outerResolve,
+			// sentReject: outerReject,
+
+			request: {
+				requestId: reqId,
+				cancel: () =>
+					this._requestQueue.splice(
```

Comment:
This can lead to never-resolved promises, when awaiting `response` in userland.

I propose we do either
1. Only allow cancelling not-yet-sent command, throw otherwise
2. When cancelling, reject the sent-but-not-yet-responded command response with a "The command was sent but then cancelled" message